### PR TITLE
(nexus-repository-migrator) Fixes Nexus 3.71 Breaking Change Upgrade

### LINKED ISSUE #2516: (nexus-repository) 3.71.* breaks upgrade
### Checklist

- [X] I have verified that this is the correct repository, and the package is maintained by the chocolatey-community user.
- [X] I have verified that this is happening in the latest available version of the package.

### Chocolatey Version

2.3.0, 2.2.2, 2.2.0, 2.1.0, 2.0.0

### Chocolatey License

Business

### Package Version

3.71.*

### Current Behaviour

Upgrades break due to the required migration of the database.

### Expected Behaviour

Upgrade should succeed.

### Steps To Reproduce

1. Have any previous version of the package installed.
2. Upgrade to the latest version.

### Environment

```markdown
- Operating System:
- PowerShell Version:
- Shell:
```

### Chocolatey Log

```shell
N/A
```

### Anything else?

This hasn't been fully filled out, more details will be added.

### PR DESCRIPTION
**Pull request by JPRuskin**
## Description

We now have a new package, `nexus-repository-migrator`, that contains logic to migrate the Nexus Repository database from 3.70.* to 3.71.*.

This new migration process:
- Checks to see if the version we're on can be migrated from (e.g. ==3.70)
- Checks to see if the install is using OrientDb (and so needs to be upgraded)
- Checks to see if the machine has sufficient memory and disk space
- Backs up the OrientDb files from the $DataDir\db directory
- Downloads the migration tool
- Migrates the backed up databases
- Copies the migrated file(s) back to the database directory
- Updates the Nexus properties to enable the datastore

By default, the package just downloads the migrator file and adds the `ConvertFrom-NexusOrientDb` script to PATH. You can also pass `/Migration` to run it during the package installation.

After this is completed, you should be able to upgrade using the next available version of Nexus Repository - the 3.71.0.6 version will _still_ break due to the merged files!

This works back to PowerShell v3, largely due to `Compress-ArchiveCompat`.

As mentioned, we now ensure that we are removing and replacing the program directory instead of copying over / merging it with the new content, as the upgrade to JRE and Nexus broke stuff with that strategy. The changes now source all connection variables in Wait-NexusAvailability (other than Hostname) from the on-disk configuration, and automatically backs-up and restores the port and SSL configuration if they are defined in the configuration. This should result in a smoother experience for people in future upgrades, and ensure `Wait-NexusAvailability` fails less frequently on SSL enabled installations.

Finally, we re-enable the update script for Nexus Repository.

The update script for the package will only find packages that match the 3.70 to 3.71 migration, currently! This is a mindful choice that we can reconsider in the future.

## Motivation and Context

Nexus' recent update has added some fairly catastrophic breaking changes, and
necessitated a large addition to the install script for the sake of migrating
the database to the new H2 format (previously only available to pro customers)

Fixes #2516

## How Has this Been Tested?

- Tested on Windows Server 2022 (PowerShell 5), upgrading from 3.70.x, with configured SSL
- Tested on Windows Server 2012 (PowerShell 3), upgrading from 3.70.x, without SSL

## Types of changes

- ~[ ] Bug fix (non-breaking change which fixes an issue)~
- ~[ ] New feature (non-breaking change which adds functionality)~
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- ~[ ] Migrated package (a package has been migrated from another repository)~

## Checklist:
- [x] My code follows the code style of this repository.
- [x] My change requires a change to documentation (this usually means the notes in the description of a package).
- [x] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [x] I have updated the package description and it is less than 4000 characters.
- [x] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)
- [ ] The added/modified package passed install/uninstall in the Chocolatey Test Environment(https://github.com/chocolatey-community/chocolatey-test-environment/). _Note that we don't support the use of any other environment_.
- [ ] The changes only affect a single package (not including meta package).

The changes affect two packages, but this has been agreed with other members of the community to be alright.

### CODE REVIEW

**Comment by JPRuskin** [comment]
I can strip out the second commit if we prefer.

File: automatic/nexus-repository/nexus-repository.nuspec [hunk 1]
```diff
@@ -40,6 +40,12 @@ You can pass parameters as follows:
 
 ## Notes
 
+- **ATTENTION BREAKING CHANGE FOR UPGRADES FROM VERSIONS BEFORE 3.71.0.6**  
+    Nexus has upgraded the bundled version of Java, and sunset the OrientDb database type.
+    This means that you will need to migrate your database from OrientDb, if you are using it.
+    The version after 3.71.0.6 will attempt to migrate the database during the package upgrade.
```
Comment (steviecoaster) [inline automatic/nexus-repository/nexus-repository.nuspec:46]: after 3.71.0.6? 

This reads like this version won't migrate to H2, when it in fact wil.

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 [hunk 2]
```diff
@@ -47,6 +48,76 @@ if ($pp.ContainsKey("BackupSslConfig")) {
...
+  Write-Error (@(
+    "Please upgrade nexus-repository to version 3.70.1-02 before upgrading further."
+    "You can do this by running 'choco upgrade nexus-repository --version 3.70.1.2 --confirm'"
+    "For more details, please see: https://help.sonatype.com/en/orient-pre-3-70-java-8-or-11.html"
+  ) -join "`n")
+  throw "Package cannot upgrade from '$($CurrentlyInstalledVersion)' to '$($env:ChocolateyPackageVersion)'"
+} elseif ($CurrentlyInstalledVersion -and [version]($CurrentlyInstalledVersion -replace '-\d+$') -eq "3.70.1" -and (Test-NexusDatabaseType -Type "OrientDb" -DataDir $TargetDataFolder -ProgramDir $TargetFolder)) {
+  Write-Host "Preparing for database migration..."
+  $MigrationFiles = Join-Path $ExtractFolder "dbmigration"
+  if ((Get-PSDrive $MigrationFiles[0]).Free -lt ($RequiredSpace = [Math]::Max((3 * (Get-ChildItem -Path $TargetDataFolder\nexus3\db -Recurse | Measure-Object Length -Sum).Sum), 10GB))) {
```
Comment (steviecoaster) [inline automatic/nexus-repository/tools/chocolateyinstall.ps1:68]: We may want to break this down a bit so it's easier to follow what's goin on. It works, but it is a bit of a symbol salad right now.

**Review by steviecoaster — DISMISSED** [review]

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 (hunk 2 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:68]: I did think about pushing this all out into a Test-migrationrequired function. It would be moving the problem, but maybe good for those who read the installer scripts!

File: automatic/nexus-repository/nexus-repository.nuspec (hunk 1 above)
Comment (JPRuskin) [reply automatic/nexus-repository/nexus-repository.nuspec:46]: To be clear, version 3.71.0.6 will never have the upgrade logic in. :(

Comment (steviecoaster) [reply automatic/nexus-repository/nexus-repository.nuspec:46]: Ooooh right, as that's already hit the feed approved. Derp. Ignore me.

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 (hunk 2 above)
Comment (steviecoaster) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:68]: Someone still has to review the code though whether it is here directly or in helper. Pick a spot, and use more variable for readability. We don't have to pay for them :)

File: automatic/nexus-repository/tools/helpers.ps1 [hunk 3]
```diff
@@ -105,3 +97,189 @@ function Wait-NexusAvailability {
...
+  }
+  try {
+    if (-not (Test-Path $DestinationPath) -and $PSCmdlet.ShouldProcess($DestinationPath, "Ensuring Folder")) {
+      $null = New-Item -Path $DestinationPath -ItemType Directory -Force
+    }
+    $PreviousProgress, $ProgressPreference = $ProgressPreference, "SilentlyContinue"
+    foreach ($Database in $IncludedDatabases) {
+      Push-Location (Join-Path $DataDir "nexus3/db/$Database")
+      $ArchiveArgs = @{
+        Path = Get-ChildItem -Path (Join-Path $DataDir "nexus3/db/$Database/*") -Exclude "cache.stt" -File
```
Comment (TheCakeIsNaOH) [inline automatic/nexus-repository/tools/helpers.ps1:219]: My understanding is that to ensure the database in a consistent state, the Nexus task to export the database should be used. I'm just concerned that this could be creating a backup with the database not working correctly.

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 [hunk 4]
```diff
@@ -35,18 +42,89 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
...
+    FileFullPath = Join-Path $toolsDir "nexus-db-migrator.jar"
+    Checksum = "02ea88e88e7d5d7f9d8a1738c9c0363fde636fc22e295efddf9d0f5e7bed982a"
+    ChecksumType = "SHA256"
+  }
+  Get-ChocolateyWebFile @MigratorDownload
+
+  try {
+    Push-Location $MigrationFiles
+    Write-Host "Migrating database from 'OrientDb' to 'H2'"
+    $JavaPath = Join-Path $TargetFolder "jre/bin/java.exe"
```
Comment (TheCakeIsNaOH) [inline automatic/nexus-repository/tools/chocolateyinstall.ps1:98]: Is this java binary always going to be java 8/11, i.e. a compatible version with the migrator `.jar`?

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 [hunk 5]
```diff
@@ -35,18 +42,89 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
...
+      "There is only $($Memory.Capacity/1GB)GB available. For more details, please see: "
+      "https://help.sonatype.com/en/orient-3-70-java-8-or-11.html#considerations-before-migrating-252166"
+    ) -join "`n")
+  }
+  New-NexusOrientDbBackup -DataDir $TargetDataFolder -ServiceName $ServiceName -DestinationPath $MigrationFiles
+
+  # See: https://help.sonatype.com/en/orientdb-downloads.html
+  $MigratorDownload = @{
+    PackageName = $env:ChocolateyPackageName
+    Url = 'https://download.sonatype.com/nexus/nxrm3-migrator/nexus-db-migrator-3.70.1-03.jar'
```
Comment (TheCakeIsNaOH) [inline automatic/nexus-repository/tools/chocolateyinstall.ps1:88]: Is it worth breaking this migrator out into a separate package so it can be updated separately, since the download page mentions that download urls for older version(s) will be removed if/when it gets updated.

File: automatic/nexus-repository/tools/helpers.ps1 (hunk 3 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/helpers.ps1:219]: That's why I'm shutting down the service.

To be clear, I have both:
- Used BeyondCompare to check that my backups are file-identical (the only difference is the compression algo used)
- Tested a _bunch_ of times that migrations succeed, and import all of the data (that I can see) successfully

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 (hunk 4 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:98]: Yes, the bundled JRE version in the Chocolatey package is always, on versions <3.71, going to be 8/11.

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 (hunk 2 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:68]: Ah, I misunderstood your complaint.

**Comment by JPRuskin** [comment]
My one fun question is "do we want to handle trying to fix up installs where they have attempted to upgrade to 3.71.0.6 and it's failed"? 

I suspect they will be in a fun state.

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 (hunk 5 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:75]: I'm going back and forth between "we should continue to support migration for several versions, so this makes good sense" (as we'd have to consider adding the migration tool to the update script in either case) and "unless we plan to support migrating indefinitely, and will consequently have to continue to upgrade this tool and test mirgations, maybe we should just say 'the upgrade path requires you to go to 3.70.1, then 3.71.x, and then it's all good'".

image

I dislike this broken chain.

File: automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1 [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    .\ConvertFrom-NexusOrientDb.ps1 -JavaPath .\path\to\java.exe
+#>
+[CmdletBinding(SupportsShouldProcess)]
+param(
+  [string]$JavaPath,
+
+  [switch]$Force
+)
+
+Import-Module $PSScriptRoot\nexus-helpers.ps1 -Verbose:$false
```
Comment (steviecoaster) [inline automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1:18]: Where is this file?

Comment (JPRuskin) [reply automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1:18]: The nuspec pulls the helpers file from nexus repository. I had written it to use it as psm1, but it turned out that NuGet treats file extension changes as a folder instead.

File: automatic/nexus-repository/tools/helpers.ps1 (hunk 3 above)
Comment (JPRuskin) [reply automatic/nexus-repository/tools/helpers.ps1:288]: Resolving this off, as I hope I answered the question / addressed the concern. If not, please let me know.

**Comment by JPRuskin** [comment]
Just to keep people updated:

I'm (we're) happy with the code in this PR, so it's ready for review. However, we're keeping it in draft because Sonatype have a few changes they're making to the migrator to not break indexes and such. Until that new version is released, we [probably] shouldn't release this package. It's arguable.

That said, please do let loose with feedback.

**Review by Windos — COMMENTED** [review]
Neither a solid approve or change request from me, just the result of my personal testing.

As mentioned in the inline comments, I originally didn't see that the migrator package had a parameter that needed to be passed in if I wanted to attempt the migration at install time. This led to issues with my trying to run the migrator after the fact.

I mentioned at least one variable defined in the install which is then used in the migration script (I didn't see any others) but even when manually re-creating that variable the migration did not succeed.

I ended up having to force a re-install of the migration package with the `/AttemptMigration` parameter and then the migration worked as expected.

I'm not sure if the best way to stop people like me from holding this wrong is in the documentation (i.e. adding to the "you need to install this migration package" error) or code (making attempting the migration the default behaviour), that is probably worth a discussion?

**Review by pauby — CHANGES_REQUESTED** [review]

File: automatic/nexus-repository/tools/chocolateyinstall.ps1 [hunk 7]
```diff
@@ -34,12 +42,21 @@ if (Test-Path "$env:ProgramFiles\nexus\bin") {
   throw "Previous version of Nexus 3 installed by setup.exe is present, please uninstall before running this package."
 }
 
+if (Test-NexusMigratorRequired -DataDir $TargetDataFolder -ProgramDir $TargetFolder) {
+  Write-Error (@(
+    "This upgrade will fail if you do not migrate your database from OrientDb before proceeding."
+    "You can do this with the nexus-repository-migrator package, or following the upgrade instructions."
```
[the remark this answers is not in the document]
Comment (JPRuskin) [reply automatic/nexus-repository/tools/chocolateyinstall.ps1:48]: I was steered away from automating the actions too much in the last community package discussion, so I ended up defaulting to just providing the file(s). I've done an attempt at improving the messaging.

File: automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1 [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      Write-Host "Migrating database from 'OrientDb' to 'H2'"
+      # We could adjust the memory values to work on more systems,
+      # but these are the recommended arguments from Sonatype.
+      $JavaArgs = @(
+        "-Xmx16G"
+        "-Xms16G"
+        "-XX:+UseG1GC"
+        "-XX:MaxDirectMemorySize=28672M"
+        "-jar"
+        $MigratorDownload.FileFullPath
```
[the remark this answers is not in the document]
Comment (JPRuskin) [reply automatic/nexus-repository-migrator/tools/ConvertFrom-NexusOrientDb.ps1:74]: Embarrassing! I used to have a fully separate download block in this script, but removed it when I realised I'd have to keep two checksums up to date 😅 

Fixed this up, great catch, thank you!

File: automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+You can either install the package and run the migrator by following the instructions in Notes, or run the script `ConvertFrom-NexusOrientDb` in the tools directory.
+
+If you have upgraded to the `nexus-repository` package version 3.71.0.6 and your installation is failing to run, you can run the `Repair-Nexus371FailedUpgrade` script in the tools directory.
+
+## Package Parameters
+
+This package supports the following parameters:
+
+* `/AttemptMigration` - Attempts to migrate your existing Sonatype Nexus install from OrientDb to a H2 database.
```
[the remark this answers is not in the document]
Comment (JPRuskin) [reply automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec:29]: > That sounds like 'We-will-give-this-a-try-fingers-crossed'

I mean, it was written by me 😅 That said, it will very much _attempt_ - it'll check a bunch of stuff first, and fail before calling the migrator if (for instance) you have less than the recommended amount of memory or disk space. I'll adjust the wording, in any case, to sound more hopeful.

**Review by pauby — CHANGES_REQUESTED** [review]
Note I'm not reviewing the code or testing the install / upgrade / uninstall.

File: automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec (hunk 9 above)
[the remark this answers is not in the document]
Comment (JPRuskin) [reply automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec]: The parameter is now `/Migrate`?

[the remark this answers is not in the document]
Comment (pauby) [reply automatic/nexus-repository-migrator/nexus-repository-migrator.nuspec]: It is. The text still says 'attempt':

> `/Migrate` - Attempts to migrate your existing Sonatype Nexus install from OrientDb to a H2 database.

**Review by pauby — DISMISSED** [review]
I know I've said this a few times, but just being clear again, I have not checked the code or tested this package. Just the wording.

**Comment by Windos** [comment]
I can re-do my testing of the code in the morning, should have results up ~16 hours

**Comment by pauby** [comment]
@TheCakeIsNaOH as you reviewed this previously, are you able to review this again and approve or add comments?

Unless anybody else feels strongly, once we have @Windos and @TheCakeIsNaOH review, I suggest we merge this.

File: automatic/nexus-repository/tools/helpers.ps1 [hunk 10]
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
...
-    $nexusPath = (Get-Content $NexusConfigFile | Where-Object {
-        $_ -match "nexus-context-path"
-      }).Split("=")[-1]
-  } else {
-    Write-Warning "Expected Nexus Config file '$($NexusConfigFile)' is not present."
-    $nexusPath, $nexusPort = '/', $Port
-  }
-
-  $NexusUri = "$($nexusScheme)://$($hostname):$($nexusPort)$($nexusPath)"
+  $NexusUri = Get-NexusUri -Hostname $Hostname
```
Comment (Windos) [inline automatic/nexus-repository/tools/helpers.ps1:153]:
```suggestion
  $NexusUri = Get-NexusUri -HostnameOverride $Hostname
```

Just wondering if this should be changed? Not sure if PowerShell is doing magic to make `-Hostname` match `HostnameOverride`

File: automatic/nexus-repository/tools/helpers.ps1 [hunk 11]
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
...
+        if ($KeyToolOutput -join "`n" -match "(?smi)Certificate\[1\]:\nOwner: CN=(?<Domain>.+?)\n") {
+          $Matches.Domain
+        }
+      } | Receive-Job -Wait
+    }
   }
 }
 
-function Wait-NexusAvailability {
+function Get-NexusUri {
```
Comment (Windos) [inline automatic/nexus-repository/tools/helpers.ps1:69]: I had issues with this function. The cert on my Nexus server is a wild card, so it doesn't have a common name or anything on it that is _specific_ to the server nor that would match the given FQDN that I pass in as a package parameter.

Also, despite my given said FQDN, it ended up using the hostname. My _expectation_ as a user is if I provide an FQDN, then that should be used (regardless of what the cert may say... if I've provided it and I got it wrong, then it's my own fault for shooting myself in the foot (imho)).

File: automatic/nexus-repository/tools/helpers.ps1 [hunk 12]
```diff
@@ -39,58 +38,122 @@ function Restore-NexusSSL {
...
+          ($script:OverriddenDomains[$CertDomain] = $HostnameOverride)
+        } elseif ($CertDomain -match '^\*') {
+          while ($script:OverriddenDomains[$CertDomain] -notlike $CertDomain) {
+            $script:OverriddenDomains[$CertDomain] = Read-Host "Please provide the FQDN for Nexus matching the '$($CertDomain)' certificate"
+          }
+          $script:OverriddenDomains[$CertDomain]
+        }
+      } else {
+        Write-Warning "Could not figure out SSL configuration for $($env:ComputerName)"
+        $env:ComputerName
```
Comment (Windos) [inline automatic/nexus-repository/tools/helpers.ps1:108]: I would _personally_ prefer this to default to localhost rather than computer name. Some environments (mine 😅) aren't going to resolve the computer name to something useful for the purposes of ensuring Nexus started.

**Review by Windos — COMMENTED** [review]
Ok, so, I tested this against a testing restore of a "real" Nexus instance that had an OrientDB. This Nexus instance uses S3 storage to host the configured repos. Despite not having access to this storage, prior to the upgrade Nexus started up fine. After the upgrade, Nexus is now no longer starting up and I can't see a _specific_ reason for this in the logs. I actually think it was to do with a NuGet v2 feed (and so it was a Nexus issue and not an upgrade issue).

Chasing that hunch, I did a take 2 and removed the suspect repo before going through the migration, and it worked!
image

I did run into some other rough edges which I've commented directly on the code itself. The result of those rough edges is that at the end of the DB migration and Nexus upgrade was it couldn't detect that Nexus was back up and running (when it absolutely was.)
image

tl;dr the migration and upgrade _work_, but in my case the rough edge of "not using the FQDN I provided" resulted in the package install/upgrades to "fail" due to not being able to validate that Nexus web interface was running.

**Review by TheCakeIsNaOH — COMMENTED** [review]
I'm not seeing anything glaringly wrong, but as I don't run Nexus on Windows, there may be something I missed.

File: automatic/nexus-repository/tools/helpers.ps1 [hunk 13]
```diff
@@ -102,6 +172,359 @@ function Wait-NexusAvailability {
...
+      # Returns if a migration is required, based on the existing version and database.
+  #>
+  [OutputType([Nullable[bool]])]
+  [CmdletBinding()]
+  param(
+    # The installed version of Nexus
+    $CurrentVersion = (Get-NexusVersion -ProgramDir $ProgramDir),
+
+    # The only version of Nexus we should be upgrading from
+    $RequiredVersion = "3.70.1",
```
Comment (Windos) [inline automatic/nexus-repository/tools/helpers.ps1:384]: Just one thing, there have been patches beyond this which will still need to be migrated from when going up to 3.71 and beyond. e.g. My Nexus instance is currently 3.70.3.1.

My suggestion would be to make this a Minimum Required Version (so test that the installed version is equal to or above without going into the 3.71.

**Review by Windos — CHANGES_REQUESTED** [review]
Other than my one comment, this all went fine and all the code looks good to me. I did test this in practice (with a test restore of a real Nexus instance).

**Comment by pauby** [comment]
@JPRuskin can you look at the last comment left and update the PR?

**Review by AdmiringWorm — DISMISSED** [review]
LGTM, however due to lack of knowledge about nexus I would defer to @Windos to give the final approval before we can merge this PR in.

**Comment by Windos** [comment]
@JPRuskin - Is this ready for a "full" re-review?

**Comment by JPRuskin** [comment]
> @JPRuskin - Is this ready for a "full" re-review?

Yes, please. Seems to work fine (other than the error output from the Nexus command, which I think may be [this](https://github.com/chocolatey/choco/issues/1016)) for both clean install and upgrade from the latest CCR version.

The build is failing, and I can't immediately see why. I am unsure if it's the error-stream stuff (as it happens directly after) but the message shows that it can't start the service. I'm not seeing that issue locally.

**Comment by Windos** [comment]
@JPRuskin the install/upgrade of the nexus-repository package is currently failing (at least for me and in AppVeyor).

The result of both the install and upgrade (from 3.71.0.601) is the same:

```text
PS C:\tmp> choco install nexus-repository -s .
Chocolatey v2.6.0
Installing the following packages:
nexus-repository
By installing, you accept licenses for the packages.
Downloading package from source 'C:\tmp'

nexus-repository v3.91.1.4
nexus-repository package files install completed. Performing other installation steps.
The package nexus-repository wants to run 'chocolateyinstall.ps1'.
Note: If you don't run this script, the installation will fail.
Note: To confirm automatically next time, use '-y' or consider:
choco feature enable -n allowGlobalConfirmation
Do you want to run the script?([Y]es/[A]ll scripts/[N]o/[P]rint): a

Downloading nexus-repository 64 bit
  from 'https://download.sonatype.com/nexus/3/nexus-3.91.1-04-win-x86_64.zip'
Progress: 100% - Completed download of C:\Users\Windos\AppData\Local\Temp\chocolatey\nexus-repository\3.91.1.4\nexus-3.91.1-04-win-x86_64.zip (439.7 MB).
Download of nexus-3.91.1-04-win-x86_64.zip (439.7 MB) completed.
Hashes match.
Extracting C:\Users\Windos\AppData\Local\Temp\chocolatey\nexus-repository\3.91.1.4\nexus-3.91.1-04-win-x86_64.zip to C:\Users\Windos\AppData\Local\Temp\chocolatey\NexusExtract...
C:\Users\Windos\AppData\Local\Temp\chocolatey\NexusExtract
Copying files to 'C:\ProgramData\nexus' with overwrite
[2026-05-11 12:07:36] [info]  ( prunsrv.c:2084) [  344] Apache Commons Daemon procrun (1.4.1.0 64-bit) started.
[2026-05-11 12:07:36] [debug] ( prunsrv.c:774 ) [  344] Installing service...
[2026-05-11 12:07:36] [info]  ( prunsrv.c:831 ) [  344] Installing service 'nexus' name 'Sonatype Nexus Repository'.
[2026-05-11 12:07:36] [debug] ( prunsrv.c:860 ) [  344] Setting service description 'Sonatype Nexus Repository'.
[2026-05-11 12:07:36] [debug] ( prunsrv.c:865 ) [  344] Setting service user 'LocalSystem'.
[2026-05-11 12:07:36] [info]  ( prunsrv.c:882 ) [  344] Service 'nexus' installed.
[2026-05-11 12:07:36] [info]  ( prunsrv.c:2190) [  344] Apache Commons Daemon procrun finished.
0
ERROR: The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Failed to start service 'Sonatype Nexus Repository (nexus)'.
The install of nexus-repository was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\nexus-repository\tools\chocolateyinstall.ps1'.
 See log for details.

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - nexus-repository (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\nexus-repository\tools\chocolateyinstall.ps1'.
 See log for details.
```

Full chocolatey.log:  [chocolatey.log](https://github.com/user-attachments/files/27575987/chocolatey.log)
